File: modules/widget-bridge/src/WidgetBridgeModule.ts
import { requireNativeModule } from "expo-modules-core";

interface WidgetBridgeModuleType {
  setWidgetData(json: string): Promise<boolean>;
  reloadAllTimelines(): Promise<boolean>;
  updateAllWidgets(): Promise<boolean>;
  getAppGroupContainerPath(): string | null;
}

// Try to load the native module
let nativeModule: WidgetBridgeModuleType | null = null;
try {
  nativeModule = requireNativeModule<WidgetBridgeModuleType>("WidgetBridge");
} catch {
  console.warn("[WidgetBridge] Native module not available");
}

export default nativeModule;

File: apps/app/modules/widget-bridge/ios/WidgetBridgeModule.swift
import ExpoModulesCore
#if canImport(WidgetKit)
import WidgetKit
#endif

public class WidgetBridgeModule: Module {
  private let appGroupId = "group.com.betazeta.jot.widgets"
  private let widgetDataKey = "countdownWidgets"

  public func definition() -> ModuleDefinition {
    Name("WidgetBridge")

    AsyncFunction("setWidgetData") { (json: String) -> Bool in
      guard let sharedDefaults = UserDefaults(suiteName: self.appGroupId) else {
        return false
      }
      sharedDefaults.set(json, forKey: self.widgetDataKey)
      sharedDefaults.synchronize()
      return true
    }

    AsyncFunction("reloadAllTimelines") { () -> Bool in
      #if canImport(WidgetKit)
      if #available(iOS 14.0, *) {
        WidgetCenter.shared.reloadAllTimelines()
        return true
      }
      #endif
      return false
    }

    AsyncFunction("updateAllWidgets") { () -> Bool in
      return false
    }

    Function("getAppGroupContainerPath") { () -> String? in
      return FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: self.appGroupId)?.path
    }
  }
}
